numpy/core/src/multiarray/ctors.c

```diff
@@ -3049,12 +3065,19 @@ _calc_length(PyObject *start, PyObject *stop, PyObject *step, PyObject **next, i
         }
         return -1;
     }
+
+    next_is_nonzero = PyObject_RichCompareBool(*next, zero, Py_NE);
```

Comment:
Could use `PyObject_IsTrue`, but that might be worse on unusual types

Reply:
@eric-wieser IMO, `PyObject_RichCompareBool` is better because this is clear what we do.